src/buf_writer.c

```diff
@@ -42,6 +43,23 @@ buf_writer_assert(buf_writer_t *buf_writer) {
 		assert(buf_writer->private_cbopaque == NULL);
 		assert(buf_writer->buf_size == 0);
 	}
+	assert(buf_writer->buf_end <= buf_writer->buf_size);
+	if (buf_writer->allow_auto_flush) {
+		assert(buf_writer->checkpoint == 0);
+		assert(!buf_writer->overflowed);
+	} else {
+		assert(buf_writer->checkpoint <= buf_writer->buf_end);
+	}
+}
+
+static void
+buf_writer_set_allow_auto_flush_internal(buf_writer_t *buf_writer,
+    bool allow_auto_flush) {
+	buf_writer->allow_auto_flush = allow_auto_flush;
+	buf_writer->checkpoint = 0;
+	if (allow_auto_flush) {
+		buf_writer->overflowed = false;
```

Comment:
Should be better to also clear overflowed; having (overflowed == true && checkpoint == 0) can cause confusion as well.

Reply:
I thought about that but decided not to. `buf_writer_set_allow_auto_flush()` should do minimal work in addition to literally setting `allow_auto_flush`, so that I can force the caller side code to have a higher code quality, i.e. to exactly stick to the cycle of `buf_writer_set_checkpoint()` -> write something -> `buf_writer_overflowed()` -> handle overflow.

In particular, the only chance this `overflowed == true && checkpoint == 0` condition can cause an issue is:
- the user first calls `buf_writer_set_allow_auto_flush(false)` when `allow_auto_flush` is already `false`, and
- the user tries to write something without first setting a checkpoint.

Both these two parts should be prohibited.

Reply:
Upon a second thought, I think when `buf_writer_set_allow_auto_flush(false)` is called, we should not clear either `overflowed` or `checkpoint`. It should just do what the name literally suggest: set `allow_auto_flush`, with minimal additional work to make things work. In other words, calling it again on a buffered writer already with `allow_auto_flush = false` should intuitively behave like a no-op, and should not change any internal state of the buffered writer.

If the caller is not careful, which means that the caller made both mistakes listed in my previous comment, then there's actually nothing dangerous:
- If `overflowed` has already been set, then the caller cannot write out anything, which is safe.
- If `overflowed` has not been set, then the buffer has not overflowed yet, so further writing to it, even if it triggers overflow, is also safe.

And on a higher level, the safety is guaranteed by that the second `buf_writer_set_allow_auto_flush(false)` behaves as a no-op so any subsequent writing simply behaves like continuation of the previous writing.

Reply:
This seems like an argument for asserting that overflowed is false when we turn on auto dumping. It's never correct not to handle overflow if you've turned off auto-dump then caused it, right?

Reply:
You mean the case where:
- `allow_auto_flush` is turned off, and then
- the writing causes overflow, and then
- `allow_auto_flush` is turned back on without first trying to handle the overflow?

In the current design this is allowed. The handling for overflow is manual and is up to the caller. The caller may want to flush when an overflow takes place, but the caller do not have to. In particular, in the case of the last-N dumping, when overflow happens, the caller may either flush, or switch `allow_auto_flush` on without flushing (though we should avoid the scenario leading to the latter decision - see my comments).